File: backend/app/models/task.py
"""
异步任务模型（简易版）

Phase 2-3 阶段实现 Producer-Consumer 任务队列的 Producer 端。
Worker 进程将在 Phase 3 实现，当前仅创建和管理 Task 记录。

任务生命周期:
    pending → running → completed
                     → failed → (手动) pending（重试）

与现有项目的关系:
    - Photo.processed_tasks (JSON) 记录每张照片已完成的分析类型
    - Task 表记录每次分析任务的状态和结果
    - 两个字段配合使用：Task 追踪过程，processed_tasks 缓存结果
"""

import uuid
import enum
from datetime import datetime
from sqlalchemy import (
    Column, String, ForeignKey, DateTime, Text, JSON,
    Enum as SAEnum, Index,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship
from app.database.session import Base


class TaskStatus(str, enum.Enum):
    pending = "pending"
    running = "running"
    completed = "completed"
    failed = "failed"


class TaskType(str, enum.Enum):
    face_detect = "face_detect"              # 人脸检测 + 特征提取
    face_cluster = "face_cluster"            # 全量人脸聚类
    image_description = "image_description"  # AI 画面描述
    image_embedding = "image_embedding"      # CLIP 向量嵌入
    quality_assessment = "quality_assessment" # 美观度/回忆价值评分
    thumbnail_generate = "thumbnail_generate" # 缩略图生成
    exif_extract = "exif_extract"            # EXIF 元数据提取
    dedup_check = "dedup_check"              # 重复照片检测


class Task(Base):
    """异步任务记录"""
    __tablename__ = "tasks"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    photo_id = Column(
        UUID(as_uuid=True), ForeignKey("photos.id", ondelete="CASCADE"),
        nullable=True, index=True, comment="关联的照片 ID（某些任务可能不关联照片）"
    )
    owner_id = Column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"),
        nullable=False, index=True, comment="任务所属用户"
    )
    task_type = Column(
        SAEnum(TaskType), nullable=False, index=True, comment="任务类型"
    )
    status = Column(
        SAEnum(TaskStatus), default=TaskStatus.pending, nullable=False, index=True,
        comment="任务状态"
    )
    progress = Column(JSON, default=dict, comment="任务进度（如 {current: 10, total: 50}）")
    result = Column(JSON, default=dict, comment="任务结果数据")
    error_message = Column(Text, nullable=True, comment="失败原因")
    created_at = Column(DateTime, default=datetime.now, comment="创建时间")
    updated_at = Column(DateTime, default=datetime.now, onupdate=datetime.now, comment="更新时间")

    __table_args__ = (
        Index("idx_tasks_status_created", "status", "created_at"),
        Index("idx_tasks_photo_type", "photo_id", "task_type"),
    )

File: frontend/src/types/task.ts
/** 后台任务状态 */
export type TaskStatus = 'pending' | 'running' | 'completed' | 'failed'

/** 任务项（对应后端 TaskResponse） */
export interface TaskItem {
  id: string
  photo_id?: string | null
  task_type: string
  status: TaskStatus
  progress?: Record<string, unknown>
  result?: Record<string, unknown>
  error_message?: string | null
  created_at: string
  updated_at?: string | null
}

/** 任务统计（对应后端 TaskStatsResponse） */
export interface TaskStats {
  total: number
  pending: number
  running: number
  completed: number
  failed: number
}

/** 任务类型中文标签（对应后端 TaskType 枚举） */
export const TASK_TYPE_LABELS: Record<string, string> = {
  face_detect: '人脸检测',
  face_cluster: '人脸聚类',
  image_description: '画面描述',
  image_embedding: '向量嵌入',
  quality_assessment: '质量评分',
  thumbnail_generate: '缩略图生成',
  exif_extract: 'EXIF 提取',
  dedup_check: '重复检测',
  object_detection: '目标检测',
}

type TagType = 'primary' | 'success' | 'info' | 'warning' | 'danger'

/** 任务状态展示元数据（标签文案 + el-tag 类型） */
export const TASK_STATUS_META: Record<TaskStatus, { label: string; type: TagType }> = {
  pending: { label: '等待中', type: 'info' },
  running: { label: '执行中', type: 'primary' },
  completed: { label: '已完成', type: 'success' },
  failed: { label: '失败', type: 'danger' },
}

/** 重新分析对话框可选的任务类型 */
export const REANALYZE_TASK_OPTIONS: { label: string; value: string }[] = [
  { label: '画面描述', value: 'image_description' },
  { label: '向量嵌入', value: 'image_embedding' },
  { label: '质量评分', value: 'quality_assessment' },
  { label: '人脸检测', value: 'face_detect' },
  { label: '目标检测', value: 'object_detection' },
  { label: 'EXIF 提取', value: 'exif_extract' },
]
